File: src/utils/curve.ts
import type { CurvePoint } from "../backend/types.ts";

// Mirrors the Rust Fritsch-Carlson monotone cubic used for the LUT.
export function evaluateCurve(curve: CurvePoint[], x: number): number {
    if (curve.length === 0) return x;
    const pts = [...curve].sort((a, b) => a.x - b.x);
    const n = pts.length;

    const first = pts[0];
    const last = pts[n - 1];
    if (first === undefined || last === undefined) return x;
    if (x <= first.x) return first.y;
    if (x >= last.x) return last.y;

    if (n === 2) {
        const t = (x - first.x) / (last.x - first.x);
        return first.y + t * (last.y - first.y);
    }

    const d: number[] = [];
    for (let i = 0; i < n - 1; i++) {
        const a = pts[i];
        const b = pts[i + 1];
        if (a === undefined || b === undefined) return x;
        d.push((b.y - a.y) / (b.x - a.x));
    }

    const m: number[] = new Array(n);
    const d0 = d[0];
    const dn = d[n - 2];
    if (d0 === undefined || dn === undefined) return x;
    m[0] = d0;
    m[n - 1] = dn;
    for (let i = 1; i < n - 1; i++) {
        const dl = d[i - 1];
        const dr = d[i];
        if (dl === undefined || dr === undefined) return x;
        m[i] = (dl + dr) / 2;
    }

    for (let i = 0; i < n - 1; i++) {
        const di = d[i];
        const mi = m[i];
        const mi1 = m[i + 1];
        if (di === undefined || mi === undefined || mi1 === undefined) return x;
        if (Math.abs(di) < 1e-12) { m[i] = 0; m[i + 1] = 0; }
        else {
            const alpha = mi / di, beta = mi1 / di;
            const h = Math.sqrt(alpha * alpha + beta * beta);
            if (h > 3) { m[i] = 3 * di * alpha / h; m[i + 1] = 3 * di * beta / h; }
        }
    }

    let lo = 0, hi = n - 2;
    while (lo < hi) {
        const mid = (lo + hi) >> 1;
        const midPt = pts[mid + 1];
        if (midPt === undefined) return x;
        if (midPt.x < x) lo = mid + 1; else hi = mid;
    }

    const i = lo;
    const pa = pts[i];
    const pb = pts[i + 1];
    const ma = m[i];
    const mb = m[i + 1];
    if (pa === undefined || pb === undefined || ma === undefined || mb === undefined) return x;
    const h = pb.x - pa.x;
    const t = (x - pa.x) / h;
    const t2 = t * t, t3 = t2 * t;
    return (2 * t3 - 3 * t2 + 1) * pa.y + (t3 - 2 * t2 + t) * h * ma
        + (-2 * t3 + 3 * t2) * pb.y + (t3 - t2) * h * mb;
}

File: core/src/color/curve.rs
use crate::types;

/// True when the curve is the identity (no effect needed).
pub fn is_linear_curve(curve: &[types::CurvePoint]) -> bool {
    curve.iter().all(|p| (p.x - p.y).abs() < 0.5)
}

/// Build a 256-entry LUT from control points using Fritsch-Carlson
/// monotone cubic interpolation.
pub fn build_curve_lut(curve: &[types::CurvePoint]) -> [u8; 256] {
    let mut lut = [0u8; 256];

    if curve.is_empty() {
        for i in 0..256 {
            lut[i] = i as u8;
        }
        return lut;
    }

    // Sort by x, deduplicate adjacent x values.
    let mut pts: Vec<(f64, f64)> = curve.iter().map(|p| (p.x, p.y)).collect();
    pts.sort_by(|a, b| a.0.partial_cmp(&b.0).unwrap_or(std::cmp::Ordering::Equal));
    pts.dedup_by(|a, b| (a.0 - b.0).abs() < 0.5);
    let n = pts.len();

    if n == 1 {
        let v = pts[0].1.round().clamp(0.0, 255.0) as u8;
        lut.fill(v);
        return lut;
    }

    // Secant slopes.
    let d: Vec<f64> = (0..n - 1)
        .map(|i| (pts[i + 1].1 - pts[i].1) / (pts[i + 1].0 - pts[i].0))
        .collect();

    // Initial tangents (Catmull-Rom start/end, average interior).
    let mut m = vec![0f64; n];
    m[0] = d[0];
    m[n - 1] = d[n - 2];
    for i in 1..n - 1 {
        m[i] = (d[i - 1] + d[i]) / 2.0;
    }

    // Fritsch-Carlson monotonicity adjustment.
    for i in 0..n - 1 {
        if d[i].abs() < 1e-12 {
            m[i] = 0.0;
            m[i + 1] = 0.0;
        } else {
            let alpha = m[i] / d[i];
            let beta = m[i + 1] / d[i];
            let h = (alpha * alpha + beta * beta).sqrt();
            if h > 3.0 {
                m[i] = 3.0 * d[i] * alpha / h;
                m[i + 1] = 3.0 * d[i] * beta / h;
            }
        }
    }

    // Evaluate at each integer 0..=255.
    for xi in 0usize..256 {
        let x = xi as f64;
        let y = if x <= pts[0].0 {
            pts[0].1
        } else if x >= pts[n - 1].0 {
            pts[n - 1].1
        } else {
            // Binary-search for the containing segment.
            let mut lo = 0usize;
            let mut hi = n - 2;
            while lo < hi {
                let mid = (lo + hi) / 2;
                if pts[mid + 1].0 < x {
                    lo = mid + 1;
                } else {
                    hi = mid;
                }
            }
            let i = lo;
            let h = pts[i + 1].0 - pts[i].0;
            let t = (x - pts[i].0) / h;
            let t2 = t * t;
            let t3 = t2 * t;
            // Cubic Hermite basis.
            (2.0 * t3 - 3.0 * t2 + 1.0) * pts[i].1
                + (t3 - 2.0 * t2 + t) * h * m[i]
                + (-2.0 * t3 + 3.0 * t2) * pts[i + 1].1
                + (t3 - t2) * h * m[i + 1]
        };
        lut[xi] = y.round().clamp(0.0, 255.0) as u8;
    }
    lut
}
